File: Programmers/Algorithm_Kit/08_dfs_bfs/002.cpp
#include <string>
#include <vector>
#include <iostream>
using namespace std;

bool visited[201];
vector<int> adj[201];

void dfs(int here) {
    visited[here] = true;

    for (int there : adj[here]) {
        if (!visited[there]) {
            dfs(there);
        }
    }

    return;
}

int solution(int n, vector<vector<int>> computers) {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL); cout.tie(NULL);

    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            if (computers[i][j] == 1) {
                adj[i].push_back(j);
                adj[j].push_back(i);
            }
        }
    }

    int answer = 0;

    for (int i = 0; i < n; i++) {
        if (!visited[i]) {
            dfs(i);
            answer++;
        }
    }

    return answer;
}

File: programmers/algorithmkit/08_dfs_bfs/002.java
import java.util.*;

class Solution {
    List<Integer>[] adj = new ArrayList[201];
    boolean[] visited = new boolean[201];
    
    public int solution(int n, int[][] computers) {
        int answer = 0;
        
        for(int i = 1; i <= n; i++) {
            adj[i] = new ArrayList<>();
        }
        
        for(int i = 0; i < n; i++) {
            for(int j = 0; j < n; j++) {
                if(computers[i][j] == 1) {
                    adj[i + 1].add(j + 1);
                    adj[j + 1].add(i + 1);
                }
            }
        }
        
        for(int i = 1; i <= n; i++) {
            if(!visited[i]) {
                dfs(i);
                answer++;
            }
        }
        
        return answer;
    }
    
    void dfs(int here) {
        visited[here] = true;
        
        for(int there : adj[here]) {
            if(!visited[there]) {
                dfs(there);
            }
        }
        
        return;
    }
}
